drgn/helpers/linux/fs.py

```diff
@@ -219,6 +219,32 @@ def dentry_path(dentry: Object) -> bytes:
     return b"/".join(reversed(components))
 
 
+def dentry_path_first_mount(dentry: Object) -> bytes:
+    """
+    Return the full path of a dentry from the root of its filesystem,
+    arbitrarily selecting the first mountpoint of its superblock.
+
+    The Linux kernel uses ``list_add_tail()`` when creating new mountpoints. So,
+    the first mount will be the oldest and thus it is likely the most relevant
+    or most "expected" mount point, as opposed to bind-mounts or mounts from
+    non-init filesystem namespaces. In short, this is very likely to get you a
+    useful filesystem path for diagnostics, without needing to have a ``struct
+    vfsmount *``.
+
+    :param dentry: ``struct dentry *``
+    """
+    # Defined as a macro in the kernel source. Introduced in 2.6.34 and has not
+    # been modified since.
+    MNT_INTERNAL = 0x4000
```

Comment:
Can you remind me what `MNT_INTERNAL` means?

Reply:
That's actually a really good question. I know that before I added this check, there were cases where I got rather unexpected results. That said, reading through the code, it should be set on superblocks which have `SB_KERNMOUNT` set, which are created from the `kern_mount()` function (rather than by userspace).

I'll see if I can go through and find the particular vmcore that led me to add this, and try to understand the context better. As it is, I've treated it a bit like magic which is less than ideal :sweat_smile:

Reply:
Ok, so I decided to go the exploratory route rather because reading the code was confusing me a bit. I made a bit more progress this way. I wrote a drgn script to find any super block with a `MNT_INTERNAL` and then printed all the mounts for that super.

It seems like a lot of kernel filesystems (e.g. selinuxfs, mqueue, tracefs, debugfs) have their first mount on the list set `MNT_INTERNAL`, and these mount points appear to be on their own root. I guess the reason is because these filesystems would exist even if userspace never mounted them, so the kernel makes its own "mount point" first in an internal root filesystem?

And I think this justifies skipping over these mounts in general. Since they're internal to the kernel, they're not seen by userspace, and so they would be a bit confusing to see when debugging. And because they are added to the list first, they will always be the first element of the list for these in-kernel filesystems.

A practical example of an incorrect value avoided by this check might be `/sys/kernel/tracing/trace_pipe`. Without the check, the function would return `b'/trace_pipe'` rather than the more useful `b'/sys/kernel/tracing/trace_pipe'` (or wherever tracefs happens to be mounted).

The flip side is that, if you managed to get your hands on a dentry which _only_ has an internal mount, then this function currently raises an error. So arguably the first internal mount should be saved, and returned as a fallback for the case where no mount is found at all.

https://gist.github.com/brenns10/1d6262133b64615c349722e143dfb3f3